### PR TITLE
chore: updating devcontainer to use new recommended settings from microsoft

### PR DESCRIPTION
**Pull request by abhilash-kumar-nair**
This PR helps resolves circular imports when specifying type hints for methods. The TYPE_CHECKING flag used is set to True by mypy  while performing type check.
Also fixed the devcontainer based on Emil's cookie cutter template to fix the python interpreter selection issues.

### CODE REVIEW

File: modelon/impact/client/jupyterhub/authorize.py [hunk 1]
```diff
@@ -1,22 +1,31 @@
...
+from modelon.impact.client.sal.context import Context
+from modelon.impact.client.sal.uri import URI
 
 logger = logging.getLogger(__name__)
 
 
-def _get_jupyter_token(credential_manager, interactive, context=None):
+def _get_jupyter_token(
+    credential_manager: CredentialManager, interactive: bool
+) -> Optional[str]:
```
Comment (efredriksson-modelon) [inline modelon/impact/client/jupyterhub/authorize.py:14]: This one should not return optional

File: modelon/impact/client/jupyterhub/authorize.py [hunk 2]
```diff
@@ -26,25 +35,25 @@ def authorize(uri, interactive, context=None, credential_manager=None, service=N
...
-            user = service.get_user_data(uri, context)
+            jupyter_context.token = credential_manager.get_key_from_prompt()
+            user = service.get_user_data(uri, jupyter_context)
         else:
             raise
 
-    if interactive:
-        credential_manager.write_key_to_file(context.token)
+    if interactive and jupyter_context.token:
+        credential_manager.write_key_to_file(jupyter_context.token)
```
Comment (efredriksson-modelon) [inline modelon/impact/client/jupyterhub/authorize.py:56]: It seems like pyright is not smart enough to get this check, maybe we could create a separate variable like:
```
jupyter_token = jupyter_context.token
```
to help it?

File: modelon/impact/client/experiment_definition/extension.py [hunk 3]
```diff
@@ -231,7 +235,7 @@ def to_dict(self):
             simulate_ext.to_dict()
 
         """
-        ext_dict = {}
+        ext_dict: Dict[str, Any] = {}
```
Comment (efredriksson-modelon) [inline modelon/impact/client/experiment_definition/extension.py:238]: This seems to result in a pyright error bellow on line 259, not sure why it gives an error or if this is correct but would be nice if we could fix it.

**Review by efredriksson-modelon — CHANGES_REQUESTED** [review]
Nice work :+1, just found some minor type issues from pyright.

**Comment by abhilash-kumar-nair** [comment]
Thanks @efredriksson-modelon  for the review! I have now fixed the pyright warnings. Could you please take a quick look? Thanks! :)

File: modelon/impact/client/jupyterhub/authorize.py [hunk 4]
```diff
@@ -37,23 +35,24 @@ def authorize(
     )
     jupyter_context = sal.JupyterContext(base=context)
     service = service or sal.JupyterHubService()
+    jupyter_token = jupyter_context.token
 
     try:
-        jupyter_context.token = _get_jupyter_token(credential_manager, interactive)
+        jupyter_token = _get_jupyter_token(credential_manager, interactive)
```
Comment (efredriksson-modelon) [inline modelon/impact/client/jupyterhub/authorize.py:41]: We still need to set the token for the jupyter_context, assigning the variable would have to be after we have done this.

File: modelon/impact/client/experiment_definition/extension.py [hunk 5]
```diff
@@ -237,7 +237,8 @@ def to_dict(self) -> Dict[str, Any]:
         """
         ext_dict: Dict[str, Any] = {}
         if self._variable_modifiers:
-            ext_dict["modifiers"] = {"variables": self._variable_modifiers}
+            ext_dict.setdefault("modifiers", {})
+            ext_dict["modifiers"]["variables"]: self._variable_modifiers
```
Comment (efredriksson-modelon) [inline modelon/impact/client/experiment_definition/extension.py:241]: typo

**Review by efredriksson-modelon — CHANGES_REQUESTED** [review]

**Review by efredriksson-modelon — APPROVED** [review]
:+1:

**Pull request merged**